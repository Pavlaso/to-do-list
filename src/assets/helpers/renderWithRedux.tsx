import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { createReduxStore, RootState } from "../../redux/store"

export const renderWithRedux = (component: any, ititialState: any = {}) => {
    const store = createReduxStore(ititialState)
    return render(
        <Provider store={store}>
            {component}
        </Provider>
    )
}