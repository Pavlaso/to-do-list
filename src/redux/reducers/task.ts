import { createSlice, PayloadAction } from "@reduxjs/toolkit";

const initialState: taskArrayType  = {
    taskArray: []
}

export const task = createSlice({
    name: 'task',
    initialState,
    reducers: {
        addTask: (state, actoin: PayloadAction<addTaskType>) => {
            state.taskArray.push({
                text: actoin.payload.text,
                folderIndex: actoin.payload.folderIndex,
                task: false,
                isWorking: true
            })
        },

        stopWorking: (state, ation: PayloadAction<number>) => {
            state.taskArray[ation.payload].isWorking = !state.taskArray[ation.payload].isWorking
        },
        deleteTask: (state, action: PayloadAction<number>) => {
            state.taskArray.splice(action.payload, 1)
        },
        markTask: (state, action: PayloadAction<number>) => {
            state.taskArray[action.payload].task = !state.taskArray[action.payload].task
        }
    }
})

export const { addTask, deleteTask, markTask, stopWorking } = task.actions

export default task.reducer

export type taskObjectType = {
    text: string
    task: boolean
    folderIndex: number
    isWorking: boolean
}

export type taskArrayType = { taskArray: Array<taskObjectType> }

export type addTaskType = {
    folderIndex: number
    text: string
}