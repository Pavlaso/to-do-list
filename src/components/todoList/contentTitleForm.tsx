import { ChangeEvent, FC } from "react"
import { folderObjectType } from "../../redux/reducers/folder"

export const ContentTitleForm: FC<ContentTitleFormType> = ({index, handleSubmit, folderArray, handleNameChange}) => {
    return (
    <form onSubmit={handleSubmit}>
        <input type="text" defaultValue={folderArray[index].name} onChange={handleNameChange} className="content__title-input"/>
        <button type="submit">âœ”</button>
    </form>
    )
}
type ContentTitleFormType = {
    index: number
    handleSubmit: (e: any) => void 
    handleNameChange: (e: ChangeEvent<HTMLInputElement>) => void 
    folderArray: folderObjectType[]
}