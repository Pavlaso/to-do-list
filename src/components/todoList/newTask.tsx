import { FC, useState } from "react";
import { useForm } from "react-hook-form";
import { useAppDispatch } from "../../assets/hooks/dispatch-selector.hook";
import { addTask } from "../../redux/reducers/task";
import plus from '../../assets/images/plus.png'



export const NewTask: FC<newTaskType> = ({index}) => {
  const dispatch = useAppDispatch()
  const [openForm, setOpenForm] = useState(false)
  const { register, handleSubmit } = useForm();

  const onSubmit = (data: dataType) => {
    dispatch(addTask({text: data.text, folderIndex: index}))
    
    setOpenForm(false)
  }

  return <div className="content__task-create">
    {!openForm 
    ? <div className="content__task-create">
        <img src={plus} alt="plus" className="content__task-create-img" />
        <button className="content__task-btn" onClick={() => setOpenForm(true)}>Новая задача</button>
      </div>
    : <form className="content__form" onSubmit={handleSubmit(onSubmit)}>
        <input className="content__crate-input" type="text" {...register("text")} placeholder='Текст задачи'/>
        <button className="content__crate-btn" type='submit'>Добавит задачу</button>
        <button className="content__crate-cancel" onClick={() => setOpenForm(false)}>Отмена</button>
      </form>
    }
  </div>
}

type dataType = {
  text: string
  folderIndex: number
}
export type newTaskType = {
  index: number
}