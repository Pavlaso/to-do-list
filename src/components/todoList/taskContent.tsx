import classNames from 'classnames'
import { FC } from 'react'
import { useAppDispatch } from '../../assets/hooks/dispatch-selector.hook'

import cross from '../../assets/images/cross.svg'
import { deleteTask, markTask, taskObjectType } from '../../redux/reducers/task'



export const TaskContent: FC<TaskContentType> = ({setId, item, num, taskArray}) => {
    const dispatch = useAppDispatch()

    const deleteTaskFunc = (index: number ) => dispatch(deleteTask(index))

    const setActiveTask = (index: number) => {
        dispatch(markTask(index))
        setId(index)
    }

    const usualStyle = (num: number) => `content__task-text${taskArray[num].task ? '_active' : ''}`
    const checboxStyle = (num: number) => classNames(
        usualStyle(num), 
        { ['content__task-notworking-text']: !taskArray[num].isWorking }
    )

    return (
        <div className="content__task" >
            <div className="content__task-container">
                <input className="content__task-checkbox" type="checkbox" />
                <div className={checboxStyle(num)} onClick={() => setActiveTask(num) }> 
                    {item.text} 
                </div>
            </div>
            { 
                !taskArray[num].isWorking && 
                <img className='content__cross' src={cross} alt="cross" onClick={() => deleteTaskFunc(num)}/> 
            }
        </div>
    )
}

type TaskContentType = {
    setId: (id: number) => void
    taskArray: taskObjectType[]
    item: taskObjectType
    num: number

}