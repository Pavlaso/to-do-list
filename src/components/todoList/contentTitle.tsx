import { ChangeEvent, FC, FormEvent } from "react"
import { useAppSelector } from "../../assets/hooks/dispatch-selector.hook"
import change from '../../assets/images/change.png'
import { getFolderArray } from "../../redux/selectors/getFolderArray"
import { ContentTitleForm } from "./contentTitleForm"


export const ContentTitle: FC<ContentTitleType> = ({index, activePopup,setActivePopup, handleSubmit, handleNameChange}) => {

    const folderArray = useAppSelector(getFolderArray)

    const titleColor = `${folderArray[index].color} content__title`

    return <div className="content__title-container">
        {
            !activePopup 
            ? <h1 className={titleColor}> 
                {folderArray[index].name}
            </h1>
            : <ContentTitleForm  
                index={index} 
                handleSubmit={handleSubmit} 
                folderArray={folderArray} 
                handleNameChange={handleNameChange} 
            />
        }
        <img src={change} alt="change-btn" className="content__title-img" onClick={() => setActivePopup(true)}/>
    </div>
}

type ContentTitleType = {
    index: number
    activePopup: boolean
    setActivePopup: (val: boolean) => void 
    handleSubmit: (e: FormEvent<HTMLFormElement>) => void 
    handleNameChange: (e: ChangeEvent<HTMLInputElement>) => void 
}