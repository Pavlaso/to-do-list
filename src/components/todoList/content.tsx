import menu_media from '../../assets/images/menu-media.png'

import cn from "classnames"

import { useAppDispatch } from '../../assets/hooks/dispatch-selector.hook'
import { changeNameFolder } from '../../redux/reducers/folder'
import { stopWorking  } from '../../redux/reducers/task'
import { ChangeEvent, FC, FormEvent, useEffect, useState } from 'react'
import { AsideContainer } from '../aside/asideContainer'
import { ContentTitle } from './contentTitle'
import { Tasks } from './tasks'

export const Content: FC<ContentType> = ({ index, setIndex }) => {
    const dispatch = useAppDispatch()

    const [activePopup, setActivePopup] = useState(false)
    const [inputState, setInputState] = useState('')
    const [activeAside, setActiveAside] = useState(false)
    const [id, setId] = useState<number | null>(null)

    const contentStyle = cn('content', {['menu-madia__content']: activeAside})

    const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => setInputState( e.target.value ) 

    const handleSubmit = (e: FormEvent<HTMLFormElement>)  =>  {
        dispatch(changeNameFolder({index, name: inputState})) 
        e.preventDefault();
        setActivePopup(false)
    }

    useEffect(() =>  {
        if(id !== null) {
            setTimeout(() => {
                dispatch(stopWorking(id))
            }, 5000)
        }   
    }, [id])

    return <div className={contentStyle}>
        
        <img className="menu-madia" src={menu_media} alt="menu media" onClick={() => setActiveAside(pre => !pre)}/>

        {
            activeAside && <AsideContainer setIndex={setIndex} />
        }

        { 
            (index === null)  
            ? <div className='content__empty'> Задачи отсутствуют </div>
            : <div>
                <ContentTitle 
                    index={index} 
                    activePopup={activePopup} 
                    setActivePopup={setActivePopup} 
                    handleSubmit={handleSubmit} 
                    handleNameChange={handleNameChange} 
                />
                <Tasks 
                    index={index} 
                    setId={setId} 
                />
            </div>
        }

    </div>
}

type ContentType = {
    index: number
    setIndex: (index: number | null) => void
}