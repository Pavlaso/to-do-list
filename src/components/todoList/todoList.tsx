import { useState } from 'react'
import { Aside } from '../aside/aside'
import { Content } from './content'

export const TodoList = () => {

    const [index, setIndex] = useState<number | null>(null)

    return <div className="body">
    <div className="container">
      <div className="body__container">
        <Aside setIndex={setIndex} />
        <Content index={index as number} setIndex={setIndex}/>
      </div>
    </div>
  </div>
}