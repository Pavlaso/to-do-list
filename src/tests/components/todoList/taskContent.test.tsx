import { getByTestId, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithRedux } from "../../../assets/helpers/renderWithRedux";
import { TaskContent } from "../../../components/todoList/taskContent";
import { taskObjectType } from "../../../redux/reducers/task";

describe("taskContent", () => {
    let task: taskObjectType;
    beforeEach(() => {
        task = {
            text: 'str',
            task: false,
            folderIndex: 0,
            isWorking: false
        }
    })
    it('Task text redered', () => {
        renderWithRedux(<TaskContent setId={jest.fn()} item={task} num={0} taskArray={[task]}/>)
        expect(screen.getByText(/str/i)).toBeInTheDocument()
        expect(screen.getByAltText(/cross/i)).toBeInTheDocument()
    })
    it('setActiveTask works', () => {
        renderWithRedux(<TaskContent setId={jest.fn()} item={task} num={0} taskArray={[task]}/>)
        userEvent.click(screen.getByRole('checkbox'))
        expect(screen.getByRole('checkbox')).toBeChecked()
    })
    it('deleteTaskFunc works', () => {
        renderWithRedux(<TaskContent setId={jest.fn()} item={task} num={0} taskArray={[task]}/>)
        const cross = screen.getByAltText(/cross/i)

        userEvent.click(cross)
        expect(cross.onclick).toBeTruthy()
    })
    it('taskContent snapshot', () => {
        const taskElem = renderWithRedux(<TaskContent setId={jest.fn()} item={task} num={0} taskArray={[task]}/>)
        expect(taskElem).toMatchSnapshot()
    })
})