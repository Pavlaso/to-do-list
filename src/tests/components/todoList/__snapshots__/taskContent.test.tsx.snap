// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`taskContent taskContent snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="content__task"
      >
        <div
          class="content__task-container"
        >
          <input
            class="content__task-checkbox"
            type="checkbox"
          />
          <div
            class="content__task-text content__task-notworking-text"
          >
            str
          </div>
        </div>
        <img
          alt="cross"
          class="content__cross"
          src="cross.svg"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="content__task"
    >
      <div
        class="content__task-container"
      >
        <input
          class="content__task-checkbox"
          type="checkbox"
        />
        <div
          class="content__task-text content__task-notworking-text"
        >
          str
        </div>
      </div>
      <img
        alt="cross"
        class="content__cross"
        src="cross.svg"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
